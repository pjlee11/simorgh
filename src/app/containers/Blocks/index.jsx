import React from 'react';
import { objectOf, arrayOf, func, shape, string, any } from 'prop-types';
import nanoid from 'nanoid';

// Inlined as this is a temporary component
const BlockString = props => {
  const stringProps = JSON.stringify(props);
  return <p>{stringProps}</p>;
};

const Blocks = ({ blocks, componentsToRender }) =>
  blocks.map((block, index) => {
    const { type, model } = block;

    const { type: typeOfPreviousBlock } = blocks[index - 1] || {};

    const Block = componentsToRender[type] || BlockString;

    return (
      <Block
        key={nanoid()}
        type={type}
        typeOfPreviousBlock={typeOfPreviousBlock}
        {...model}
      />
    );
  });

Blocks.propTypes = {
  blocks: arrayOf(
    shape({
      type: string.isRequired,
      model: objectOf(any).isRequired,
    }),
  ).isRequired,
  componentsToRender: objectOf(func).isRequired,
};

export default Blocks;
