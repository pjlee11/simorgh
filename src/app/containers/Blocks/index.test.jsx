import React from 'react';
import { shouldShallowMatchSnapshot } from '../../helpers/tests/testHelpers';
import Blocks from './index';
import { blockContainingText, singleTextBlock } from '../../models/blocks';

describe('Blocks', () => {
  const blocks = [
    blockContainingText('headline', 'This is a headline!'),
    blockContainingText('subheadline', 'This is a subheadline!'),
    singleTextBlock('This is some text content!'),
    {
      type: 'test',
      model: {
        blocks: [
          {
            model: {
              text: 'This is some test content!',
            },
          },
        ],
      },
    },
    {
      type: 'image',
      model: {
        blocks: [
          {
            type: 'rawImage',
            model: {
              width: 640,
              height: 420,
              locator: 'E7DB/production/_101655395_paulineclayton.jpg',
              originCode: 'cpsprodpb',
              copyrightHolder: 'BBC',
            },
          },
          blockContainingText('altText', 'Pauline Clayton'),
          blockContainingText(
            'caption',
            'Former embroider Pauline Clayton described the gift as "lovely"',
          ),
        ],
      },
    },
  ];

  const componentsToRender = {
    headline: () => <h1>A headline</h1>,
    subheadline: () => <h2>A subheadline</h2>,
    text: () => <p>Some text</p>,
  };

  shouldShallowMatchSnapshot(
    'should render correctly',
    <Blocks blocks={blocks} componentsToRender={componentsToRender} />,
  );
});
