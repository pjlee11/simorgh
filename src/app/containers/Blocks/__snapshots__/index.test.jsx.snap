// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks should render correctly 1`] = `
Array [
  <headline
    blocks={
      Array [
        Object {
          "model": Object {
            "blocks": Array [
              Object {
                "model": Object {
                  "text": "This is a headline!",
                },
                "type": "paragraph",
              },
            ],
          },
          "type": "text",
        },
      ]
    }
    type="headline"
  />,
  <subheadline
    blocks={
      Array [
        Object {
          "model": Object {
            "blocks": Array [
              Object {
                "model": Object {
                  "text": "This is a subheadline!",
                },
                "type": "paragraph",
              },
            ],
          },
          "type": "text",
        },
      ]
    }
    type="subheadline"
    typeOfPreviousBlock="headline"
  />,
  <text
    blocks={
      Array [
        Object {
          "model": Object {
            "text": "This is some text content!",
          },
          "type": "paragraph",
        },
      ]
    }
    type="text"
    typeOfPreviousBlock="subheadline"
  />,
  <BlockString
    blocks={
      Array [
        Object {
          "model": Object {
            "text": "This is some test content!",
          },
        },
      ]
    }
    type="test"
    typeOfPreviousBlock="text"
  />,
  <BlockString
    blocks={
      Array [
        Object {
          "model": Object {
            "copyrightHolder": "BBC",
            "height": 420,
            "locator": "E7DB/production/_101655395_paulineclayton.jpg",
            "originCode": "cpsprodpb",
            "width": 640,
          },
          "type": "rawImage",
        },
        Object {
          "model": Object {
            "blocks": Array [
              Object {
                "model": Object {
                  "blocks": Array [
                    Object {
                      "model": Object {
                        "text": "Pauline Clayton",
                      },
                      "type": "paragraph",
                    },
                  ],
                },
                "type": "text",
              },
            ],
          },
          "type": "altText",
        },
        Object {
          "model": Object {
            "blocks": Array [
              Object {
                "model": Object {
                  "blocks": Array [
                    Object {
                      "model": Object {
                        "text": "Former embroider Pauline Clayton described the gift as \\"lovely\\"",
                      },
                      "type": "paragraph",
                    },
                  ],
                },
                "type": "text",
              },
            ],
          },
          "type": "caption",
        },
      ]
    }
    type="image"
    typeOfPreviousBlock="test"
  />,
]
`;
